### PR TITLE
Assignment 2

### PR DESCRIPTION
**Pull request by rahulsnkr**

### CODE REVIEW

**Review by vilas897 — CHANGES_REQUESTED** [review]
Great work on the assignment! Although, try making a few changes. 
Merge sort seems to be taking too much time. Just check your code again and try to optimize it. Ideally, merge sort should be the fastest among the 5. 
Also, write a non-tail recursive solution to Fibonacci, time it and include the comparison between both non-tail recursive as well as the tail recursive version of the algorithm in your readme file.

File: FP-Erlang/Week 3/Assignment 2/Rahul/assignment2.erl [hunk 1]
```diff
@@ -0,0 +1,58 @@
+-module(assignment2).
+-export([qsort/1, insort/1, bubsort/1, selsort/1, mergesort/1, unique/1, replicate/2, reverse/1, zip/2, fibonacci/1]).
+
+qsort([]) -> [];
+qsort([Pivot|T]) ->
+qsort([X||X <- T, X<Pivot]) ++ [Pivot] ++ qsort([X||X <- T, X>=Pivot]).
+
+insort([H|T]) -> insort([H],T).
+insort(Sorted,[]) -> Sorted;
+insort(Sorted,[H|T]) -> insort([X||X <- Sorted, X<H]++[H]++[X||X <- Sorted, X>=H],T).
```
Comment (grakshith) [inline FP-Erlang/Week 3/Assignment 2/Rahul/assignment2.erl:10]: Do you notice how similar insertion sort and quick sort are in functional programming?
Rather, all sorts are similar in functional languages. These names like quick sort, insertion sort are given to algorithms implemented in imperative languages like C, C++, Java

File: FP-Erlang/Week 3/Assignment 2/Rahul/assignment2.erl [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+bubsort([], [], Sorted) -> Sorted.
+
+selsort(Unsorted) -> selsort(Unsorted,[],[]).
+selsort([H|T],Passed, Sorted) -> selsort([X||X <- T , X<H],[H]++Passed++[X||X <- T , X>=H], Sorted);
+selsort([],[H|T],Sorted) -> selsort(T,[],Sorted++[H]);
+selsort([], [], Sorted) -> Sorted.
+
+mergesort(Unsorted) -> mergesort(Unsorted, []).
+mergesort([], [H|[H1|T]]) -> mergesort(H,H1,[],T,[]);
+mergesort([],[H|T]) when T== [] -> H;
```
Comment (grakshith) [inline FP-Erlang/Week 3/Assignment 2/Rahul/assignment2.erl]: Can be written as `mergesort([], [H|[]])`

File: FP-Erlang/Week 3/Assignment 2/Rahul/assignment2.erl [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+mergesort([H|T], SplitList) -> mergesort(T, SplitList++[[H]]).
+mergesort([],[],Sorted,T,Newlist) -> mergesort(Sorted,T,Newlist);
+mergesort([],L2,Sorted,T,NewList) -> mergesort([],[],Sorted++L2,T,NewList);
+mergesort(L1,[],Sorted,T,NewList) -> mergesort([],[],Sorted++L1,T,NewList);
+mergesort([H1|T1],[H2|T2],Sorted,T,NewList) when H1<H2-> mergesort(T1,[H2|T2],Sorted++[H1],T,NewList);
+mergesort([H1|T1],[H2|T2],Sorted,T,NewList) when H2=<H1-> mergesort([H1|T1],T2,Sorted++[H2],T,NewList).
+mergesort(Sorted,[],NewList) -> mergesort([], NewList++[Sorted]);
+mergesort(Sorted,[H|T],NewList) when T==[] -> mergesort(H,[],[],T,NewList++[Sorted]);
```
Comment (grakshith) [inline FP-Erlang/Week 3/Assignment 2/Rahul/assignment2.erl]: Same as the above comment. Can be written as `mergesort(Sorted,[H|[]],NewList)`

File: FP-Erlang/Week 3/Assignment 2/Rahul/assignment2.erl [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+
+zip(List1, List2) -> zip(List1, List2, []).
+zip(T1,T2,NewList) when T1==[]; T2==[] -> NewList;
+zip([H1|T1], [H2|T2], NewList) -> zip(T1,T2,NewList++[{H1,H2}]).
+
+fibonacci(1) -> [0];
+fibonacci(2) -> [0,1];
+fibonacci(N) -> fibonacci(N-2,[0,1],[0,1]).
+fibonacci(0,Series,_) -> Series;
+fibonacci(N,Series,[H|[H1|_]]) -> fibonacci(N-1,Series++[H+H1],[H1,H+H1]).
```
Comment (grakshith) [inline FP-Erlang/Week 3/Assignment 2/Rahul/assignment2.erl]: Where's the non tail recursive one?

**Review by grakshith — CHANGES_REQUESTED** [review]

File: FP-Erlang/Week 3/Assignment 2/Rahul/assignment2.erl (hunk 1 above)
Comment (rahulsnkr) [reply FP-Erlang/Week 3/Assignment 2/Rahul/assignment2.erl:10]: Yeah I did notice it, I actually thought I wrote the same code twice quite a few times.

**Comment by grakshith** [comment]
Why did you close this?

**Review by vilas897 — APPROVED** [review]

**Pull request merged**